File: NexusGameSDK.cs
/**
 * NEXUS Game SDK - Unity (C#)
 * Production-grade client for leaderboard, telemetry, crashes, feature flags, A/B testing
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.Profiling;

public class NexusGameSDK : MonoBehaviour
{
    private static NexusGameSDK _instance;
    private string _backendUrl = "http://127.0.0.1:7000";
    private string _playerId;
    private bool _initialized = false;
    private Queue<Action> _eventQueue = new Queue<Action>();
    private Dictionary<string, object> _sessionMeta = new Dictionary<string, object>();
    private float _lastHeartbeat;
    private const float HeartbeatInterval = 30f;

    public static NexusGameSDK Instance => _instance ??= FindObjectOfType<NexusGameSDK>();

    public void Initialize(string playerId, string backendUrl = null)
    {
        if (_initialized) return;

        _playerId = playerId;
        if (backendUrl != null) _backendUrl = backendUrl;

        DontDestroyOnLoad(gameObject);
        _initialized = true;

        _sessionMeta["device_model"] = SystemInfo.deviceModel;
        _sessionMeta["gpu_name"] = SystemInfo.graphicsDeviceName;
        _sessionMeta["os"] = SystemInfo.operatingSystem;
        _sessionMeta["ram_mb"] = SystemInfo.systemMemorySize;
        _sessionMeta["gpu_memory_mb"] = SystemInfo.graphicsMemorySize;

        Debug.Log($"[NEXUS] Initialized for player {_playerId}");
        StartCoroutine(HeartbeatLoop());
    }

    private IEnumerator HeartbeatLoop()
    {
        while (_initialized)
        {
            yield return new WaitForSeconds(HeartbeatInterval);
            SendTelemetry("heartbeat", new Dictionary<string, object>
            {
                { "fps", (int)(1f / Time.deltaTime) },
                { "heap_mb", (int)(Profiler.GetTotalMemoryUsed() / 1024 / 1024) }
            });
        }
    }

    public void SubmitScore(int score, Dictionary<string, object> metadata = null)
    {
        var data = new Dictionary<string, object>
        {
            { "player_id", _playerId },
            { "score", score },
            { "meta", metadata ?? new Dictionary<string, object>() }
        };
        StartCoroutine(PostRequest("/leaderboard/submit", data));
    }

    public void SendTelemetry(string eventName, Dictionary<string, object> payload = null)
    {
        var data = new Dictionary<string, object>
        {
            { "player_id", _playerId },
            { "event", eventName },
            { "payload", payload ?? new Dictionary<string, object>() }
        };
        StartCoroutine(PostRequest("/telemetry", data));
    }

    public void ReportCrash(string stackTrace, Dictionary<string, object> deviceInfo = null)
    {
        var data = new Dictionary<string, object>
        {
            { "player_id", _playerId },
            { "stacktrace", stackTrace },
            { "device_info", deviceInfo ?? _sessionMeta }
        };
        StartCoroutine(PostRequest("/crash", data));
    }

    public void GetLeaderboard(int limit, System.Action<List<Dictionary<string, object>>> callback)
    {
        StartCoroutine(GetRequest($"/leaderboard/top?limit={limit}", callback));
    }

    public void GetFeatureFlags(System.Action<Dictionary<string, object>> callback)
    {
        StartCoroutine(GetRequest("/feature-flags", callback));
    }

    public void GetABVariant(System.Action<string> callback)
    {
        StartCoroutine(GetRequest($"/ab-variant?player_id={_playerId}", (obj) =>
        {
            if (obj is Dictionary<string, object> dict && dict.ContainsKey("variant"))
                callback((string)dict["variant"]);
        }));
    }

    private IEnumerator PostRequest(string endpoint, Dictionary<string, object> data)
    {
        string json = JsonUtility.ToJson(new Wrapper { data = data });
        using (UnityWebRequest req = new UnityWebRequest(_backendUrl + endpoint, "POST"))
        {
            req.uploadHandler = new UploadHandlerRaw(Encoding.UTF8.GetBytes(json));
            req.downloadHandler = new DownloadHandlerBuffer();
            req.SetRequestHeader("Content-Type", "application/json");

            yield return req.SendWebRequest();

            if (req.result != UnityWebRequest.Result.Success)
                Debug.LogWarning($"[NEXUS] POST error: {req.error}");
        }
    }

    private IEnumerator GetRequest(string endpoint, System.Action<object> callback)
    {
        using (UnityWebRequest req = UnityWebRequest.Get(_backendUrl + endpoint))
        {
            yield return req.SendWebRequest();

            if (req.result == UnityWebRequest.Result.Success)
            {
                var json = req.downloadHandler.text;
                Debug.Log($"[NEXUS] GET {endpoint}: {json}");
                // Parse JSON and invoke callback
                callback(new Dictionary<string, object> { { "response", json } });
            }
            else
                Debug.LogWarning($"[NEXUS] GET error: {req.error}");
        }
    }

    [Serializable]
    private class Wrapper { public Dictionary<string, object> data; }
}

File: NexusGameSDK.cpp
// NEXUS Game SDK - Unreal Engine (C++)
// Implementation: NexusGameSDK.cpp
#include "NexusGameSDK.h"
#include "Http.h"
#include "HttpModule.h"
#include "Interfaces/IHttpResponse.h"
#include "GenericPlatform/GenericPlatformMisc.h"
#include "Misc/FileHelper.h"

ANexusGameSDK::ANexusGameSDK()
    : bInitialized(false), HeartbeatTimer(0.0f)
{
    PrimaryActorTick.TickInterval = 0.1f;
    PrimaryActorTick.bCanEverTick = true;
    HttpModule = &FHttpModule::Get();
}

void ANexusGameSDK::BeginPlay()
{
    Super::BeginPlay();
}

void ANexusGameSDK::Tick(float DeltaTime)
{
    Super::Tick(DeltaTime);
    if (bInitialized)
    {
        HeartbeatTimer += DeltaTime;
        if (HeartbeatTimer >= 30.0f)
        {
            OnHeartbeat();
            HeartbeatTimer = 0.0f;
        }
    }
}

void ANexusGameSDK::Initialize(const FString& InPlayerId, const FString& InBackendUrl)
{
    if (bInitialized) return;

    PlayerId = InPlayerId;
    BackendUrl = InBackendUrl;
    bInitialized = true;

    FString DeviceModel = FGenericPlatformMisc::GetDeviceId();
    FString OS = FPlatformMisc::GetOperatingSystemId();
    
    UE_LOG(LogTemp, Warning, TEXT("[NEXUS] Initialized for player %s"), *PlayerId);
}

void ANexusGameSDK::SubmitScore(int32 Score, const FString& MetadataJSON)
{
    if (!bInitialized) return;

    FString JsonData = FString::Printf(
        TEXT("{\"player_id\":\"%s\",\"score\":%d,\"meta\":%s}"),
        *PlayerId, Score, *MetadataJSON
    );
    PostRequest(TEXT("/leaderboard/submit"), JsonData);
}

void ANexusGameSDK::SendTelemetry(const FString& EventName, const FString& PayloadJSON)
{
    if (!bInitialized) return;

    FString JsonData = FString::Printf(
        TEXT("{\"player_id\":\"%s\",\"event\":\"%s\",\"payload\":%s}"),
        *PlayerId, *EventName, *PayloadJSON
    );
    PostRequest(TEXT("/telemetry"), JsonData);
}

void ANexusGameSDK::ReportCrash(const FString& StackTrace, const FString& DeviceInfoJSON)
{
    if (!bInitialized) return;

    FString JsonData = FString::Printf(
        TEXT("{\"player_id\":\"%s\",\"stacktrace\":\"%s\",\"device_info\":%s}"),
        *PlayerId, *StackTrace, *DeviceInfoJSON
    );
    PostRequest(TEXT("/crash"), JsonData);
}

void ANexusGameSDK::PostRequest(const FString& Endpoint, const FString& JsonData)
{
    if (!HttpModule) return;

    FHttpRequestRef Request = HttpModule->CreateRequest();
    Request->SetURL(BackendUrl + Endpoint);
    Request->SetVerb(TEXT("POST"));
    Request->SetHeader(TEXT("Content-Type"), TEXT("application/json"));
    Request->SetContentAsString(JsonData);
    Request->ProcessRequest();
}

void ANexusGameSDK::GetRequest(const FString& Endpoint, FHttpRequestCompleteDelegate Callback)
{
    if (!HttpModule) return;

    FHttpRequestRef Request = HttpModule->CreateRequest();
    Request->SetURL(BackendUrl + Endpoint);
    Request->SetVerb(TEXT("GET"));
    Request->OnProcessRequestComplete() = Callback;
    Request->ProcessRequest();
}

void ANexusGameSDK::GetLeaderboard(int32 Limit, FOnLeaderboardReceived OnComplete)
{
    FString Endpoint = FString::Printf(TEXT("/leaderboard/top?limit=%d"), Limit);
    GetRequest(Endpoint, FHttpRequestCompleteDelegate::CreateLambda(
        [OnComplete](FHttpRequestPtr Request, FHttpResponsePtr Response, bool bWasSuccessful)
        {
            if (bWasSuccessful && Response)
                OnComplete.ExecuteIfBound(true, Response->GetContentAsString());
            else
                OnComplete.ExecuteIfBound(false, TEXT(""));
        }
    ));
}

void ANexusGameSDK::GetFeatureFlags(FOnLeaderboardReceived OnComplete)
{
    GetRequest(TEXT("/feature-flags"), FHttpRequestCompleteDelegate::CreateLambda(
        [OnComplete](FHttpRequestPtr Request, FHttpResponsePtr Response, bool bWasSuccessful)
        {
            if (bWasSuccessful && Response)
                OnComplete.ExecuteIfBound(true, Response->GetContentAsString());
            else
                OnComplete.ExecuteIfBound(false, TEXT(""));
        }
    ));
}

void ANexusGameSDK::GetABVariant(FOnVariantReceived OnComplete)
{
    FString Endpoint = FString::Printf(TEXT("/ab-variant?player_id=%s"), *PlayerId);
    GetRequest(Endpoint, FHttpRequestCompleteDelegate::CreateLambda(
        [OnComplete](FHttpRequestPtr Request, FHttpResponsePtr Response, bool bWasSuccessful)
        {
            FString Variant = TEXT("");
            if (bWasSuccessful && Response)
                Variant = Response->GetContentAsString();
            OnComplete.ExecuteIfBound(bWasSuccessful, Variant);
        }
    ));
}

void ANexusGameSDK::OnHeartbeat()
{
    FString HeartbeatData = TEXT("{\"fps\":60,\"memory_mb\":512}");
    SendTelemetry(TEXT("heartbeat"), HeartbeatData);
}
